ssl/statem/extensions_srvr.c

```diff
@@ -1614,6 +1614,10 @@ EXT_RETURN tls_construct_stoc_key_share(SSL *s, WPACKET *pkt,
         }
         return EXT_RETURN_NOT_SENT;
     }
+    if ((s->ext.psk_kex_mode & TLSEXT_KEX_MODE_FLAG_KE_DHE) == 0) {
+        /* Explicitly not doing DHE; don't send key share */
+        return EXT_RETURN_NOT_SENT;
```

Comment:
Is this right? I'm not familiar with this code, but it looks like this gets filled in by `tls_parse_ctos_psk_kex_modes`, which is the client's PSK capabilities. But the server might not be using a PSK, in which case the server should still send key_share.

I imagine somewhere in the parameter selection process, there's some logic that decides:
* Is the server using a PSK?
* If so, is the server doing so with psk_dhe_ke or psk_ke?

If the answers are (yes, psk_ke), the server should omit key_share. Otherwise, the server should send key_share.

Reply:
I think you are correct and we need another check here; thanks for catching it.
This is obscured in the tests because openssl as a client always sends "psk_key_exchange_modes" even if there is no PSK (and the extra testing I did with a patched version that doesn't send psk_ke_dhe was targeted at the behavior during resumption of a different API, and so didn't test full handshakes).
Unfortunately, it looks like there isn't a great way to check for "did we use a PSK" already, so we might have to add a new field for that.

Reply:
It is possible that we can use `s->hit` to stand in for "server used PSK", but I will want to carefully check the various ways we can get a PSK before pushing that patch to the PR.  (It is set somewhat far away from the decision on whether or not to use a PSK, which makes me somewhat uneasy about relying on it.)

Reply:
Could you please fix this so it can be considered for merge into 3.0?

Reply:
Yes, this is near the top of my list again (I feel like I have been staggering from one crisis to the next...)

Reply:
> It is possible that we can use `s->hit` to stand in for "server used PSK"

(we can't.)

Reply:
Oops, spoke too soon. Actually we can, since the `psk_server_callback` and `psk_find_session_callback` are only called from `tls_parse_ctos_psk()`, and the bit I forgot was that we specifically process that extension inside `ssl_get_prev_session()`, and if the latter returns a session it counts as a 'hit'.  (If the callbacks find a session but it's not acceptable in a way that's not a fatal error, the session is not used, not associated with the SSL on return, `ssl_get_prev_session()` returns 0, and it is not a 'hit'.
I should probably tidy up my notes and put them in the tree somewhere...